### DISCUSSION TITLE
Create a custom install callback, to avoid default cmd screen on windows.

### DISCUSSION THREAD

**Discussion by mdhuzaifapatel** [dennisvang/notsotuf-example#17] [Q&A]
@dennisvang As you mentioned earlier that, 

> The progress hook displays download progress, not update progress. So, 100% progress means the update has been downloaded, but it has not been installed yet.
> 
> On windows, by default, the client runs the install script in a new command window. This is where the actual installation happens.
> This only happens once for each update. If you don't like this, you can provide a custom install callback.

So, could you please let me know, how can I create a custom callback? Do I have to make some changes in this [file](https://github.com/dennisvang/tufup/blob/6d2c707adc66fd7637739f31f9859b0298a4053d/src/tufup/utils/platform_specific.py#L210) ? Because modifying the library's source code is generally not recommended.

I'm already using tkinter to display a message box, but it is being used as a progress hook for "download progress", I want the same message box to be used for "update progress"

**Comment by dennisvang** [comment]
> So, could you please let me know, how can I create a custom callback? ...

@mdhuzaifapatel You can pass a custom `install` function into `Client.download_and_apply_update()`, like this:

```python
def custom_install(src_dir, dst_dir, purge_dst_dir=False, exclude_from_purge=None, **kwargs):
    # your custom installation steps
    ...

...
client.download_and_apply_update(install=custom_install)
...
```
Your custom install function would be a modified/simplified copy of [`_install_update_win`][1] (assuming you're on windows).

Be ***very*** careful though, as you're going to be messing with your end-user's filesystem.

For more information on using custom `install` functions, search the issues on tufup and tufup-example, looking for "custom install". 

Also, study the tufup source code in detail, to understand how everything works.

> ... I want the same message box to be used for "update progress"

This is not possible with tufup, at the moment.

[1]: https://github.com/dennisvang/tufup/blob/master/src/tufup/utils/platform_specific.py#L120

**Reply by mdhuzaifapatel** [reply]
Can you pls elaborate, I'm unable to achieve this. 
I just want to remove the cmd screen which is being displayed for few seconds when update is being  apllied.

As you mentioned earlier that "custom_install function" would be a modified/simplified copy of  "_install_update_win", so I tried writing custom_install, but I failed.
Please tell me how can write a custom_insatall, in order to remove the cmd screen. (I'm on Windows)

**Reply by dennisvang** [reply]
You probably just want to modify the value of `creationflags` on [this line][1]. 

See [subprocess docs][2] for available options.

Basically, you could just copy [_install_update_win](https://github.com/dennisvang/tufup/blob/master/src/tufup/utils/platform_specific.py#L120), modify ***only*** that line, and use it [as described above][3].

If it does not work, please post the relevant part of your code so I can point out what's wrong.

[1]: https://github.com/dennisvang/tufup/blob/6d2c707adc66fd7637739f31f9859b0298a4053d/src/tufup/utils/platform_specific.py#L210 
[2]: https://docs.python.org/3/library/subprocess.html#windows-constants
[3]: https://github.com/dennisvang/tufup-example/discussions/17#discussioncomment-7100228

**Reply by mdhuzaifapatel** [reply]
It worked, Thanks!

The default cmd screen is not displaying when update is performed.
Now, I'm working on automatic relaunch of the app as soon as the update is performed.

**Reply by mdhuzaifapatel** [reply]
Hello I need your help, everything was working fine earlier but now I'm facing one issue.

This is my code 

```python
def show_update_message():
    import tkinter as tk
    from tkinter import messagebox
    root = tk.Tk()
    root.withdraw()  # Hide the main window
    messagebox.showinfo("Update Complete", "App updated successfully! Restart your CoBOT")
    root.destroy()
```

```python
def custom_install(
        src_dir: Union[pathlib.Path, str],
        dst_dir: Union[pathlib.Path, str],
        purge_dst_dir: bool,
        exclude_from_purge: List[Union[pathlib.Path, str]],
        as_admin: bool = False,
        batch_template: str = WIN_BATCH_TEMPLATE,
        batch_template_extra_kwargs: Optional[dict] = None,
        log_file_name: Optional[str] = None,
        robocopy_options_override: Optional[List[str]] = None,
):
    if batch_template_extra_kwargs is None:
        batch_template_extra_kwargs = dict()
    # collect robocopy options
    if robocopy_options_override is None:
        robocopy_options = list(WIN_ROBOCOPY_OVERWRITE)
        if purge_dst_dir:
            robocopy_options.append(WIN_ROBOCOPY_PURGE)
            if exclude_from_purge:
                robocopy_options.append(WIN_ROBOCOPY_EXCLUDE_FROM_PURGE)
                robocopy_options.extend(exclude_from_purge)
    else:
        robocopy_options = robocopy_options_override
    options_str = ' '.join(robocopy_options)
    log_lines = ''
    if log_file_name:
        log_file_path = pathlib.Path(dst_dir) / log_file_name
        log_lines = WIN_LOG_LINES.format(log_file_path=log_file_path)
        logger.info(f'logging install script output to {log_file_path}')

    script_content = batch_template.format(
        src_dir=src_dir,
        dst_dir=dst_dir,
        robocopy_options=options_str,
        log_lines=log_lines,
        delete_self=WIN_BATCH_DELETE_SELF,
        **batch_template_extra_kwargs,
    )
    logger.debug(f'writing windows batch script:\n{script_content}')
    with NamedTemporaryFile(
            mode='w', prefix=WIN_BATCH_PREFIX, suffix=WIN_BATCH_SUFFIX, delete=False
    ) as temp_file:
        temp_file.write(script_content)
    logger.debug(f'temporary batch script created: {temp_file.name}')
    script_path = pathlib.Path(temp_file.name).resolve()
    logger.debug(f'starting script in new console: {script_path}')
    # start the script in a separate process, non-blocking
    if as_admin:
        run_bat_as_admin(file_path=script_path)
    else:
        subprocess.Popen(
            [script_path], creationflags=subprocess.CREATE_NO_WINDOW
        )
    logger.debug('exiting')
    show_update_message()
```


```python
def update():
    client = Client(
        app_name=settings.APP_NAME,
        app_install_dir=settings.INSTALL_DIR,
        current_version=settings.APP_VERSION,
        metadata_dir=settings.METADATA_DIR,
        metadata_base_url=settings.METADATA_BASE_URL,
        target_dir=settings.TARGET_DIR,
        target_base_url=settings.TARGET_BASE_URL,
        refresh_required=False,
    )

    # Perform update
    if client.check_for_updates():
        client.download_and_apply_update(
            skip_confirmation=True,
            progress_hook=progress_hook,
            purge_dst_dir=False,
            exclude_from_purge=None,
            log_file_name='install.log',
            install=custom_install
        )
        return True

    return False

```

# *********************** Tufup code ENDS ******************************


```python
if __name__ == '__main__':

    for dir_path in [settings.INSTALL_DIR, settings.METADATA_DIR, settings.TARGET_DIR]:
        dir_path.mkdir(exist_ok=True, parents=True)

    if not settings.TRUSTED_ROOT_DST.exists():
        shutil.copy(src=settings.TRUSTED_ROOT_SRC, dst=settings.TRUSTED_ROOT_DST)

    multiprocessing.freeze_support()

    processes = list(map(lambda x: (x.info['name'], x.info['pid']), psutil.process_iter(['name', 'pid'])))
    app_instances = [y for x, y in processes if x == "App.exe"]

    print("*************", app_instances)

    update_status = update()
    print(update_status)

    if update_status:
        if len(app_instances) > 0:
            print('kill')
            for instance in app_instances:
                try:
                    process = psutil.Process(instance)
                    process.terminate()  # Terminate the process
                    process.wait()  # Wait for the process to finish terminating
                except psutil.NoSuchProcess:
                    pass
        else:
            print('else')
            os.abort()

    else:
        print('no update')
        if len(app_instances) > 1:
            for instance in app_instances:
                exe_path = psutil.Process(instance).exe()
                    import tkinter
                    from tkinter import messagebox

                    top = tkinter.Tk()
                    top.geometry("150x150")
                    top.withdraw()
                    messagebox.showwarning("Message", "One app instance is already opened.")
                    os.abort()
```
                    
In this code, the update() function is returning True if the server is running but it should return True only when updates are available and applied, it is returning True even if there is no update. Can you please help me out with this? 

If update return True it looks for app_insatnces and if they are more than one it kills them, else it exits the script. This part is working fine but because update is returning True everytime a message is being displayed "update complete" even if there was no update, I want to avoid this.
 Thanks.

**Comment by dennisvang** [comment]
> In this code, the update() function is returning True if the server is running but it should return True only when updates are available and applied, it is returning True even if there is no update.

@mdhuzaifapatel  I would need to see some debug log messages to answer this. 

If you don't have a custom `logging` configuration you can use e.g. `logging.basicConfig(level=logging.DEBUG)` at the start of your app.

Then run the app and post the console output. (make sure to remove any sensitive information from the output first)

**Reply by mdhuzaifapatel** [reply]
```
2023-10-18 13:25:40,403 [DEBUG] azure_interface  __init__: Azure Services config created
************* []
2023-10-18 13:25:45,502 [INFO] app  custom_install: logging install script output to C:\App_Tufup\temp\my_app\install.log
2023-10-18 13:25:45,502 [INFO] app  custom_install: logging install script output to C:\App_Tufup\temp\my_app\install.log
2023-10-18 13:25:45,502 [DEBUG] app  custom_install: writing windows batch script:                                                  
@echo off                                                                                                                           
                                                                                                                                    
call :log > "C:\App_Tufup\temp\my_app\install.log" 2>&1                                                                  
:log                                                                                                                                
                                                                                                                                    
echo Moving app files...                                                                                                            
robocopy "C:\Users\xyz\AppData\Local\Temp\tufup" "C:\App_Tufup\temp\my_app" /e /move /v /w:2                         
echo Done.                                                                                                                          
(goto) 2>nul & del "%~f0"                                                                                                           
                                                                                                                                    
2023-10-18 13:25:45,502 [DEBUG] app  custom_install: writing windows batch script:                                                  
@echo off

call :log > "C:\App_Tufup\temp\my_app\install.log" 2>&1
:log

echo Moving app files...
robocopy "C:\Users\xyz\AppData\Local\Temp\tufup" "C:\App_Tufup\temp\my_app" /e /move /v /w:2
echo Done.
(goto) 2>nul & del "%~f0"

2023-10-18 13:25:45,502 [DEBUG] app  custom_install: temporary batch script created: C:\Users\xyz\AppData\Local\Temp\tufupx_84yvd3.bat
2023-10-18 13:25:45,502 [DEBUG] app  custom_install: temporary batch script created: C:\Users\xyz\AppData\Local\Temp\tufupx_84yvd3.bat
2023-10-18 13:25:45,502 [DEBUG] app  custom_install: starting script in new console: C:\Users\xyz\AppData\Local\Temp\tufupx_84yvd3.bat
2023-10-18 13:25:45,502 [DEBUG] app  custom_install: starting script in new console: C:\Users\xyz\AppData\Local\Temp\tufupx_84yvd3.bat
2023-10-18 13:25:45,517 [DEBUG] app  custom_install: exiting
2023-10-18 13:25:45,517 [DEBUG] app  custom_install: exiting
True
else

```

When is press 'OK' in tkinter messagebox then it prints 'else' 
i just don't want the update function to return true if there is no update available


This is server console when there is no update, if there is a update it will have tar file mentioned here
```
::1 - - [18/Oct/2023 12:19:10] "GET /metadata/timestamp.json HTTP/1.1" 200 -
::1 - - [18/Oct/2023 12:22:43] code 404, message File not found
::1 - - [18/Oct/2023 12:22:43] "GET /metadata/2.root.json HTTP/1.1" 404 -
::1 - - [18/Oct/2023 12:22:43] "GET /metadata/timestamp.json HTTP/1.1" 200 -
::1 - - [18/Oct/2023 12:23:25] code 404, message File not found
::1 - - [18/Oct/2023 12:23:25] "GET /metadata/2.root.json HTTP/1.1" 404 -
::1 - - [18/Oct/2023 12:23:25] "GET /metadata/timestamp.json HTTP/1.1" 200 -
::1 - - [18/Oct/2023 12:23:57] code 404, message File not found
::1 - - [18/Oct/2023 12:23:57] "GET /metadata/2.root.json HTTP/1.1" 404 -
::1 - - [18/Oct/2023 12:23:57] "GET /metadata/timestamp.json HTTP/1.1" 200 -
::1 - - [18/Oct/2023 12:25:05] code 404, message File not found
::1 - - [18/Oct/2023 12:25:05] "GET /metadata/2.root.json HTTP/1.1" 404 -
::1 - - [18/Oct/2023 12:25:05] "GET /metadata/timestamp.json HTTP/1.1" 200 -
::1 - - [18/Oct/2023 13:25:03] code 404, message File not found
::1 - - [18/Oct/2023 13:25:03] "GET /metadata/2.root.json HTTP/1.1" 404 -
::1 - - [18/Oct/2023 13:25:03] "GET /metadata/timestamp.json HTTP/1.1" 200 -
::1 - - [18/Oct/2023 13:25:41] code 404, message File not found
::1 - - [18/Oct/2023 13:25:41] "GET /metadata/2.root.json HTTP/1.1" 404 -   
::1 - - [18/Oct/2023 13:25:41] "GET /metadata/timestamp.json HTTP/1.1" 200 -
```

**Reply by dennisvang** [reply]
This is only part of the output, and it shows that some update is being installed fro somewhere. 

However, the relevant part is missing: I need to see all the log output from the `check_for_updates()` method.

**Comment by dennisvang** [comment]
@mdhuzaifapatel  On a side note: What are you trying to achieve with all the process handling?

Windows is not linux, you cannot properly replace an executable while it is running. That's why the original windows install function [exits the running process][1] after starting robocopy in a separate process.

[1]: https://github.com/dennisvang/tufup/blob/08ff73e904560b23a5076523a74ef1c751200a7f/src/tufup/utils/platform_specific.py#L214

**Reply by mdhuzaifapatel** [reply]
So should I add  "sys.exit(0)" in my custom_install too ?

please check the previous reply i have made some changes in it

**Reply by dennisvang** [reply]
What are you trying to achieve with all the process handling?

**Reply by mdhuzaifapatel** [reply]
I got this, but earlier the update function was not running when there was no new update (tar file) found on server, but now it is always running even if there is no new update on server

**Comment by mdhuzaifapatel** [comment]
> What are you trying to achieve with all the process handling?

I'm killing the current running processes of my app (it has some processes) , so that tufup can do it's work.

**Reply by dennisvang** [reply]
Why are there multiple processes?

**Reply by mdhuzaifapatel** [reply]
They are related to the app's features, it is written by my team.
So I was trying to kill those processes if update is available so that tufup works fine, once it worked, but now its not working.

When there are processes running its killing them, but in the case when there are no processes running, this time it is going into the else condition and app is exiting

**Reply by dennisvang** [reply]
This process juggling sounds like it is outside the scope of tufup. I cannot help with that.

Please provide the relevant logging output so I can try to figure out why your `check_for_updates()` call would be returning truthy.

**Comment by mdhuzaifapatel** [comment]
> logging.basicConfig(level=logging.DEBUG)



> This process juggling sounds like it is outside the scope of tufup. I cannot help with that.
> 
> Please provide the relevant logging output so I can try to figure out why your `check_for_updates()` call would be returning truthy.

How can I get logs for this?

**Reply by dennisvang** [reply]
Run your app on your dev system, or whatever, with logging level set to DEBUG.

**Reply by dennisvang** [reply]
Also see tufup-example.
